torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.h

```diff
@@ -1,37 +1,487 @@
+// partially mapped on the left and right side. Since this class is  intended to
+// line up "inner dimensions" of tensors through out the graph for the purpose
+// of unrolling and vectorization, it only tracks partial dimensions as they are
+// on the right hand side of iteration domains. For example in the last case we
+// would only identify tv0's 3*5*7 dimension as being a mapping with extent 7.
+// If we further had:
+//   tv0[5*7*11]
+//   tv1[5*7, 11] = view(tv0)
+//   tv2[5, 7*11] = view(tv1)
+// with tv2 and [7*11] as the reference and ids (this could be a valid example
+// from the pointwise scheduler).
+// (1) tv1 would:
+//     map on 5*7 with extent 7
+//     map on 11 with extent 11.
+// (1) tv0 would:
+//     map on 5*7*11 with size 7*11
+//
+// Finally if we have:
+// tv0[3, 5, 7]
+// tv1[7, 5, 3] = view(tv0)
+// tv2[3, 5, 7] = view(tv1)
+// with tv2 mapping on 5, 7
+// We use fractional, symbolic, and conditional mappings so tv1:
+//   maps on 3 with extent 3
+//   maps on 5 with extent 5
+//   maps on 7 with extent (5*7)/(5*3)
+// Then tv0:
+//   maps on 7 with extent 7
+//   maps on 5 with extent 5
+class TORCH_CUDA_CU_API ContiguousInnerDimensionsMapper
+    : public MaxInfoSpanningTree::Propagator {
+ public:
+  ContiguousInnerDimensionsMapper() = delete;
+
+  static ContiguousInnerDimensionsMapper map(
+      TensorView* reference,
+      std::vector<IterDomain*> ids,
+      std::shared_ptr<const ComputeAtMap> ca_map,
+      const std::unordered_set<Split*>& divisible_splits) {
+    ContiguousInnerDimensionsMapper contig_inner_mapper(
+        reference, std::move(ids), ca_map, divisible_splits);
+    MaxRootDomainInfoSpanningTree tree(reference);
+
+    tree.traverse(&contig_inner_mapper);
+    return contig_inner_mapper;
+  }
+
+  static ContiguousInnerDimensionsMapper map(
+      TensorView* reference,
+      std::vector<IterDomain*> ids) {
+    auto ca_map = std::make_shared<ComputeAtMap>(reference->fusion());
+    auto divisible_splits =
+        getAllDivisibleSplits(reference->fusion(), ca_map.get());
+    return ContiguousInnerDimensionsMapper::map(
+        reference, ids, ca_map, divisible_splits);
+  }
+
+  virtual void propagateC2P(TensorView* from, TensorView* to) override;
+  virtual void propagateP2C(TensorView* from, TensorView* to) override;
+  virtual void propagateSibling(TensorView* from, TensorView* to) override;
+
+  const std::unordered_map<TensorView*, std::vector<IterDomain*>>&
+  mappedRootIds() const {
+    return projected_root_ids_;
+  }
+
+  const std::unordered_map<TensorView*, std::vector<IterDomain*>>&
+  mappedRFactorIds() const {
+    return projected_rfactor_ids_;
+  }
+
+  ProjectedExtent& getMappedExtent(IterDomain* id) {
+    if (projected_extent_.find(id) != projected_extent_.end()) {
+      return projected_extent_.at(id);
+    }
+    ProjectedExtent init_extent;
+    init_extent.multiplyDenominatorValue(id->fusion()->oneVal());
```

Comment:
nit: Should this be done in a constructor of `ProjectedExtent`?

Reply:
Isn't necessary to initialize the Projected extent anymore.